diff --git a/pyparsing/core.py b/pyparsing/core.py
index 0d1ca6fb..bac41127 100644
--- a/pyparsing/core.py
+++ b/pyparsing/core.py
@@ -23,7 +23,6 @@
 import copy
 import warnings
 import re
-import sre_constants
 import sys
 from collections.abc import Iterable
 import traceback
@@ -2773,7 +2772,7 @@
 
             try:
                 self.re = re.compile(self.reString)
-            except sre_constants.error:
+            except re.error:
                 self.re = None
             else:
                 self.re_match = self.re.match
@@ -2932,7 +2931,7 @@
             try:
                 self.re = re.compile(self.pattern, self.flags)
                 self.reString = self.pattern
-            except sre_constants.error:
+            except re.error:
                 raise ValueError(
                     "invalid pattern ({!r}) passed to Regex".format(pattern)
                 )
@@ -3168,7 +3167,7 @@
             self.re = re.compile(self.pattern, self.flags)
             self.reString = self.pattern
             self.re_match = self.re.match
-        except sre_constants.error:
+        except re.error:
             raise ValueError(
                 "invalid pattern {!r} passed to Regex".format(self.pattern)
             )
diff --git a/pyparsing/helpers.py b/pyparsing/helpers.py
index 799de749..be8a3657 100644
--- a/pyparsing/helpers.py
+++ b/pyparsing/helpers.py
@@ -312,7 +312,7 @@ def one_of(
 
             return ret
 
-        except sre_constants.error:
+        except re.error:
             warnings.warn(
                 "Exception creating Regex for one_of, building MatchFirst", stacklevel=2
             )
